In [1]:
import pandas as pd
import numpy as np
import scipy.sparse as sps
import re
import nltk
from scipy import sparse
from nltk import bigrams    
from unicodedata import normalize
from nltk.corpus import stopwords
from collections import Counter

# Tema: Expansão de Consultas
### Autor: Luiz Fernando da Silva
Neste laboratório analizarei um conjunto de notícias do estadão armazenados em um arquivo csv utilizando técnicas de expansão de consultas para, por fim, responder as seguinstes questões:
* Quais os termos retornados para a expansão de cada consulta?
* Você acha que esses termos são de fato relacionados com a consulta original? Justifique.
* Compare os documentos retornados para a consulta original com a consulta expandida. Quais resultados você acha que melhor capturam a necessidade de informação do usuário? Por que?
* A expansão de consultas é mais adequada para melhorar o recall ou o precision? Por que?


In [2]:
dados = pd.read_csv('data/estadao_noticias_eleicao.csv')

In [3]:
def limpar_texto(texto):
    pattern = re.compile('[^a-zA-Z0-9 ]')
    texto = normalize('NFKD', texto).encode('ASCII', 'ignore').decode('ASCII')
    return pattern.sub(' ', texto)

In [4]:
words = [limpar_texto(stopword) for stopword in stopwords.words('portuguese')]

In [5]:
conteudo = dados.titulo + " " + dados.subTitulo + " " + dados.conteudo
conteudo = conteudo.fillna("")
conteudo = conteudo.apply(limpar_texto)
ids = dados.idNoticia

In [6]:
tokens = conteudo.apply(nltk.word_tokenize)
term_frequence = tokens.apply(Counter)

In [7]:
index = {}

for i in range(len(tokens)):
    id_noticia = ids[i]
    palavras = tokens[i]
    for palavra in palavras:
        palavra = palavra.lower()
        if palavra not in index:
            index[palavra] = {}
        
        id_rec = index[palavra].get(id_noticia)
        
        if not id_rec:
            docs = index[palavra]
            docs[id_noticia] = term_frequence[i][palavra]

In [8]:
def gera_docs_peso(termos):
    docs_peso = {}
    
    for i in range(len(termos)):
        termo = termos[i]
        docs = index[termo]
        for doc_id in docs:
            tf = docs[doc_id]
            
            if doc_id not in docs_peso:
                docs_peso[doc_id] = np.array([0 if j != i else tf for j in range(len(termos))])
            else:
                doc_vector = docs_peso[doc_id]
                doc_vector[i] = tf
    return docs_peso


In [9]:
def gera_query_vetor(termos):
    vetor = np.array([1 if index.get(termo) else 0 for termo in termos])
    return vetor

In [10]:
def busca(termos, gerador_query, gerador_doc_vetor):
    docs_peso = gerador_doc_vetor(termos)
    query = gerador_query(termos)
    
    doc_rank = sorted(list(docs_peso.items()), key=lambda doc: np.dot(doc[1], query), reverse=True) 
    return [doc[0] for doc in doc_rank]

In [11]:
def buscar_por_tf(termos):
    return busca(termos, gera_query_vetor, gera_docs_peso)

# Função para gerar matrix esparsa de coocorrência
Este código pode ser encontrado em [https://github.com/allansales/information-retrieval/blob/master/Lab%202/coocurrence_matrix.ipynb](https://github.com/allansales/information-retrieval/blob/master/Lab%202/coocurrence_matrix.ipynb)

In [12]:
def co_occurrence_matrix(corpus):
    vocab = set(corpus)
    vocab = list(vocab)
    n = len(vocab)
   
    vocab_to_index = {word:i for i, word in enumerate(vocab)}
    
    bi_grams = list(bigrams(corpus))

    bigram_freq = nltk.FreqDist(bi_grams).most_common(len(bi_grams))

    I=list()
    J=list()
    V=list()
    
    for bigram in bigram_freq:
        current = bigram[0][1]
        previous = bigram[0][0]
        count = bigram[1]

        I.append(vocab_to_index[previous])
        J.append(vocab_to_index[current])
        V.append(count)
        
    co_occurrence_matrix = sparse.coo_matrix((V,(I,J)), shape=(n,n))

    return co_occurrence_matrix, vocab_to_index

In [13]:
tokens_lists = conteudo.apply(lambda text: text.lower().split())

In [14]:
tokens = [token for tokens_list in tokens_lists for token in tokens_list if token not in words]

In [15]:
matrix, vocab = co_occurrence_matrix(tokens)

# Consult Bigram Frequency

In [16]:
consultable_matrix = matrix.tocsr()

In [17]:
def consult_frequency(w1, w2):
    return(consultable_matrix[vocab[w1],vocab[w2]])

In [18]:
def get_co_ocurrence(word):
    list_of_occurency = consultable_matrix[vocab[word]].getrow(0).toarray()[0]
    indexs, frequency = zip(*sorted(enumerate(list_of_occurency), key=lambda x: x[1], reverse=True))
    return indexs[:3], frequency[:3]

In [19]:
def busca_expasao(termo):
    ocurrecy = get_co_ocurrence(termo)
    expansao = [word for key in ocurrecy[0] for word in vocab.keys() if vocab[word] == key]
    expansao.append(termo)
    exp = buscar_por_tf(expansao)
    return (expansao, exp)
    

# Gerando Busca Expansiva

In [20]:
# Usado o termo Dilma
termo = 'dilma'
expansao, doc_exp = busca_expasao(termo)
busca_original = buscar_por_tf([termo])
print(expansao)
print(len(doc_exp))
print(len(busca_original))

# Usado o termo petrobras
termo = 'petrobras'
expansao, doc_exp = busca_expasao(termo)
busca_original = set(buscar_por_tf([termo]))
print(expansao)
print(len(doc_exp))
print(len(busca_original))

# Usado o termo aecio
termo = 'aecio'
expansao, doc_exp = busca_expasao(termo)
busca_original = set(buscar_por_tf([termo]))
print(expansao)
print(len(doc_exp))
print(len(busca_original))

['rousseff', 'disse', 'aecio', 'dilma']
3853
2669
['paulo', 'graca', 'disse', 'petrobras']
3886
963
['neves', 'disse', 'afirmou', 'aecio']
3658
1578


# Análise

**Tabela dos resultados das buscas originais e expandidas.**

| Termo     | Termos expanão         | Busca Original | Busca Expandida|
|---------- |:----------------------:|:--------------:| --------------:|
| Dilma     | rousseff, disse, aecio | 2669           | 3853           |
| Petrobrás | paulo, graca, disse    | 963            | 3886           |
| Aécio     | neves, disse, afirmou  | 1578           | 3658           |

* Quais os termos retornados para a expansão de cada consulta?
> Observando a tablea de resultados acima temos que os termos retornados para as buscas de Dilma, Petrobrás e Aécio são:
> * Dilma
>> * rousseff 
>> * disse 
>> * aecio
> * Petrobrás
>> * paulo 
>> * graca
>> * disse
> * Aécio
>> * neves
>> * disse
>> * afirmou


* Você acha que esses termos são de fato relacionados com a consulta original? Justifique.
> <p style="text-align: justify">Os termos retornados estão relacionados com as consultas pois, são eles quem mais aparecem em conjunto com a busca desejada e podem se gerar uma maior aproximação com que o usuário deseja pesquisar. Podemo ver com mais clareza na tabela de resultados acima, onde ao pesquisar por dilma um dos resultados dos termos de expansão é Rousseff, que é seu sobrenome. Outra análise que pode ser feita é calcular quantos dos documentos retornados pela busca original também são retornados pela busca expandida. Neste caso, mesmo não adicionado os termos da consulta original e deixando apenas os termos que mais estão correlacionados para fazer a busca, a quantidade de documentos que estão na busca original e que não estão na busca expandida é mínima, o que mostra que esses termos estão bastante relacionados aos termos da consulta original.</p>

* Compare os documentos retornados para a consulta original com a consulta expandida. Quais resultados você acha que melhor capturam a necessidade de informação do usuário? Por que?
> <p style="text-align: justify">A busca expandida retorna resultados melhores para a busca, pois como ela usa os termos que estão mais relacionados com os termos originais, fica mais expressiva e retornará também resultados de outras possíveis palavras chave que o usuário poderia usar em futuras buscas. Entretanto, quanto mais termos a busca expandida tiver, maior vai ser a quantidade de documentos retornados por ela, e se, não houver um rankeamento adequado dos resultados, o usuário pode levar mais tempo procurando os documentos que realmente são relevantes para ele.</p>

* A expansão de consultas é mais adequada para melhorar o recall ou o precision? Por que?
> <p style="text-align: justify">Expandindo a consulta o recall irá melhorar, pois ele é a razão entre o total de documentos relevantes na busca e e a quantidade total de documentos relevantes, ou seja, quanto maior for a quantidade de documentos retornados maior será a quantidade de documentos relevantes que estarão contidos nessa busca. Por outro lado, a precisão diminui já que ela é a razão entre a quantidade de documentos relevantes na busca e a quantidade de documentos totais presentes na busca. Formula do recall: $\frac{tp}{tp+fn}$; formula do precision: $\frac{tp}{tp+fp}$; onde, tp é o total de documentos relevantes na busca, fn é o total de documentos relevantes não presentes na busca e fp é o total de documentos não relevantes na busca.</p>